File: Rosterra/Models/RosterCalendar.swift
import Foundation

/// Centralised calendar/formatter configuration for all roster date math.
///
/// The web app interprets shift wall-clock times in `Australia/Adelaide`
/// (see `ROSTER_TIME_ZONE`) and starts weeks on Monday. We do the same here so
/// "today", week boundaries, and submit gates behave identically on any device.
enum RosterCalendar {
    static let timeZone = TimeZone(identifier: "Australia/Adelaide") ?? .current

    /// Gregorian calendar, Monday-first, pinned to the business timezone.
    /// Cached: `Calendar` is a value type with no mutable shared state, and it
    /// is read on virtually every date operation, so rebuilding it per access
    /// was pure waste.
    static let calendar: Calendar = {
        var cal = Calendar(identifier: .gregorian)
        cal.timeZone = timeZone
        cal.firstWeekday = 2 // Monday
        return cal
    }()

    static let dayFormatter: DateFormatter = {
        let f = DateFormatter()
        f.calendar = Calendar(identifier: .gregorian)
        f.locale = Locale(identifier: "en_US_POSIX")
        f.timeZone = timeZone
        f.dateFormat = "yyyy-MM-dd"
        return f
    }()

    /// Start of the day (in business TZ) for a given instant.
    static func startOfDay(_ date: Date) -> Date {
        calendar.startOfDay(for: date)
    }

    /// Monday 00:00 of the week containing `date`.
    static func weekStart(_ date: Date = Date()) -> Date {
        let comps = calendar.dateComponents([.yearForWeekOfYear, .weekOfYear], from: date)
        return calendar.date(from: comps) ?? startOfDay(date)
    }

    /// The `yyyy-MM-dd` key of the Monday of the week containing `date`.
    static func weekStartKey(_ date: Date = Date()) -> String {
        dayFormatter.string(from: weekStart(date))
    }

    /// The seven days (Mon...Sun) of the week containing `date`.
    static func weekDays(for date: Date) -> [Date] {
        let start = weekStart(date)
        return (0..<7).compactMap { calendar.date(byAdding: .day, value: $0, to: start) }
    }

    static func addWeeks(_ weeks: Int, to date: Date) -> Date {
        calendar.date(byAdding: .weekOfYear, value: weeks, to: date) ?? date
    }

    static func addDays(_ days: Int, to date: Date) -> Date {
        calendar.date(byAdding: .day, value: days, to: date) ?? date
    }

    static func todayKey(_ now: Date = Date()) -> String {
        dayFormatter.string(from: now)
    }

    static func dateFromKey(_ key: String) -> Date? {
        dayFormatter.date(from: key)
    }

    // MARK: - Months (payslip month filter)

    /// "yyyy-MM" key of the month containing `date` (business timezone).
    static func monthKey(_ date: Date = Date()) -> String {
        let comps = calendar.dateComponents([.year, .month], from: date)
        return monthKey(year: comps.year ?? 1970, month: comps.month ?? 1)
    }

    static func monthKey(year: Int, month: Int) -> String {
        String(format: "%04d-%02d", year, month)
    }

    /// (year, month) of a "yyyy-MM" key; nil when malformed.
    static func monthKeyComponents(_ key: String) -> (year: Int, month: Int)? {
        let parts = key.split(separator: "-")
        guard parts.count == 2, let year = Int(parts[0]), let month = Int(parts[1]),
              (1...12).contains(month) else { return nil }
        return (year, month)
    }

    /// First instant of the month for a "yyyy-MM" key (business timezone).
    static func monthStartDate(_ key: String) -> Date? {
        guard let comps = monthKeyComponents(key) else { return nil }
        return calendar.date(from: DateComponents(year: comps.year, month: comps.month, day: 1))
    }

    /// Half-open day-key bounds of a month: ("yyyy-MM-01", first day of the
    /// next month). Because `yyyy-MM-dd` sorts lexicographically — and so do
    /// document ids prefixed with it — these bounds drive string range queries.
    static func monthDayKeyBounds(_ key: String) -> (start: String, end: String)? {
        guard let comps = monthKeyComponents(key) else { return nil }
        let next = comps.month == 12 ? (comps.year + 1, 1) : (comps.year, comps.month + 1)
        return ("\(monthKey(year: comps.year, month: comps.month))-01",
                "\(monthKey(year: next.0, month: next.1))-01")
    }

    /// The "yyyy-MM" key `offset` months from `key` (e.g. -1 = previous month).
    static func monthKey(byAdding offset: Int, to key: String) -> String? {
        guard let start = monthStartDate(key),
              let shifted = calendar.date(byAdding: .month, value: offset, to: start) else { return nil }
        return monthKey(shifted)
    }
}

File: android/app/src/main/java/com/surainvestments/rosterra/core/calendar/RosterCalendar.kt
package com.surainvestments.rosterra.core.calendar

import java.time.DayOfWeek
import java.time.Instant
import java.time.LocalDate
import java.time.LocalDateTime
import java.time.LocalTime
import java.time.ZoneId
import java.time.ZonedDateTime
import java.time.format.DateTimeFormatter
import java.time.temporal.TemporalAdjusters

/**
 * Port of iOS `RosterCalendar` — Australia/Adelaide, Monday-start weeks.
 */
object RosterCalendar {
    val zoneId: ZoneId = ZoneId.of("Australia/Adelaide")
    private val dayFormatter: DateTimeFormatter = DateTimeFormatter.ofPattern("yyyy-MM-dd")
    private val monthFormatter: DateTimeFormatter = DateTimeFormatter.ofPattern("yyyy-MM")

    fun now(): ZonedDateTime = ZonedDateTime.now(zoneId)

    fun todayKey(now: ZonedDateTime = now()): String = dayKey(now.toLocalDate())

    fun dayKey(date: LocalDate): String = date.format(dayFormatter)

    fun dateFromKey(key: String): LocalDate? = runCatching { LocalDate.parse(key, dayFormatter) }.getOrNull()

    fun weekStart(date: LocalDate = now().toLocalDate()): LocalDate =
        date.with(TemporalAdjusters.previousOrSame(DayOfWeek.MONDAY))

    fun weekStart(instant: Instant): LocalDate =
        weekStart(instant.atZone(zoneId).toLocalDate())

    fun weekStartKey(date: LocalDate = now().toLocalDate()): String = dayKey(weekStart(date))

    fun weekDays(forDate: LocalDate): List<LocalDate> {
        val start = weekStart(forDate)
        return (0..6).map { start.plusDays(it.toLong()) }
    }

    fun addWeeks(weeks: Int, to: LocalDate): LocalDate = to.plusWeeks(weeks.toLong())

    fun addDays(days: Int, to: LocalDate): LocalDate = to.plusDays(days.toLong())

    fun addDays(days: Int, to: ZonedDateTime): ZonedDateTime = to.plusDays(days.toLong())

    fun isWeekend(dateKey: String): Boolean {
        val date = dateFromKey(dateKey) ?: return false
        val dow = date.dayOfWeek
        return dow == DayOfWeek.SATURDAY || dow == DayOfWeek.SUNDAY
    }

    fun monthKey(date: LocalDate = now().toLocalDate()): String = date.format(monthFormatter)

    fun monthKey(year: Int, month: Int): String = "%04d-%02d".format(year, month)

    fun monthKeyComponents(key: String): Pair<Int, Int>? {
        val parts = key.split("-")
        if (parts.size != 2) return null
        val year = parts[0].toIntOrNull() ?: return null
        val month = parts[1].toIntOrNull() ?: return null
        if (month !in 1..12) return null
        return year to month
    }

    fun monthStartDate(key: String): LocalDate? {
        val (year, month) = monthKeyComponents(key) ?: return null
        return LocalDate.of(year, month, 1)
    }

    fun monthDayKeyBounds(key: String): Pair<String, String>? {
        val (year, month) = monthKeyComponents(key) ?: return null
        val next = if (month == 12) (year + 1) to 1 else year to (month + 1)
        return "${monthKey(year, month)}-01" to "${monthKey(next.first, next.second)}-01"
    }

    fun monthKey(byAdding: Int, to: String): String? {
        val start = monthStartDate(to) ?: return null
        return monthKey(start.plusMonths(byAdding.toLong()))
    }

    fun zonedDateTime(dateKey: String, timeHHmm: String): ZonedDateTime? {
        val date = dateFromKey(dateKey) ?: return null
        val parts = timeHHmm.split(":")
        if (parts.size < 2) return null
        val hour = parts[0].toIntOrNull() ?: return null
        val minute = parts[1].toIntOrNull() ?: return null
        return ZonedDateTime.of(LocalDateTime.of(date, LocalTime.of(hour, minute)), zoneId)
    }

    fun toEpochMilli(zdt: ZonedDateTime): Long = zdt.toInstant().toEpochMilli()
}
